lib/teleterm/services/clientcache/clientcache.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+// Cache stores remote clients keyed by cluster URI.
+// Safe for concurrent access.
+// Closes all clients and wipes the cache on Close.
+type Cache struct {
+	clusters.Resolver
+	mu  sync.Mutex
+	log *logrus.Entry
+	// clients keep mapping between cluster URI
+	// (both root and leaf) and proxy clients
+	clients map[uri.ResourceURI]*client.ProxyClient
+	// group prevents duplicate requests to create remote clients
+	// for a given cluster URI
+	group singleflight.Group
+}
+
+// Config describes the client cache configuration.
+type Config struct {
+	clusters.Resolver
+	Log *logrus.Entry
+}
+
+// New creates an instance of Cache.
+func New(c Config) *Cache {
+	return &Cache{
+		log:      c.Log,
+		clients:  make(map[uri.ResourceURI]*client.ProxyClient),
+		Resolver: c.Resolver,
+	}
+}
+
+// Get returns a proxy client from the cache if there is one,
+// otherwise it dials the remote server.
+func (c *Cache) Get(ctx context.Context, clusterURI uri.ResourceURI) (*client.ProxyClient, error) {
+	groupClt, err, _ := c.group.Do(clusterURI.String(), func() (any, error) {
+		if proxyClient := c.getFromCache(clusterURI); proxyClient != nil {
+			return proxyClient, nil
+		}
+
+		_, clusterClient, err := c.ResolveCluster(clusterURI)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		//nolint:staticcheck // SA1019. TODO(gzdunek): Update to use client.ClusterClient.
+		newProxyClient, err := clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// We'll save the remote client in the cache, so we don't have to
+		// build a new connection next time.
+		// All remote clients will be closed when the daemon exits.
+		if err := c.addToCache(clusterURI, newProxyClient); err != nil {
+			c.log.WithError(err).Errorf("An error occurred while adding remote client for %q to cache.", clusterURI)
+		} else {
+			c.log.Infof("Added remote client for %q to cache.", clusterURI)
+		}
+
+		return newProxyClient, nil
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clt, ok := groupClt.(*client.ProxyClient)
+	if !ok {
+		return nil, trace.BadParameter("unexpected type %T received for proxy client", groupClt)
+	}
+
+	return clt, nil
+}
+
+// InvalidateForRootCluster closes and removes clients from the cache
+// for the root cluster and its leaf clusters.
+func (c *Cache) InvalidateForRootCluster(rootClusterURI uri.ResourceURI) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var (
+		errors  []error
+		deleted []uri.ResourceURI
+	)
+
+	for resourceURI, clt := range c.clients {
+		if resourceURI.GetRootClusterURI() == rootClusterURI {
+			if err := clt.Close(); err != nil {
+				errors = append(errors, err)
+			}
+			deleted = append(deleted, resourceURI.GetClusterURI())
+			delete(c.clients, resourceURI)
+		}
+	}
+
+	c.log.Infof("Invalidated cached remote clients for root cluster %q: %v", rootClusterURI, deleted)
```

Comment:
As per https://github.com/gravitational/teleport/blob/master/rfd/0154-logging-guidelines.md#logging-standards :
> Embrace structured logging. Prefer using static messages and attributes instead of relying on fmt.Sprintf style formatting. Enrich log messages with additional attributes to provide additional context.

Reply:
Thanks, TIL about `c.log.WithField`.